### PR TITLE
[Northamptonshire] add streetlight category message

### PR DESCRIPTION
**Pull request by struan**
Fixes mysociety/fixmystreet-commercial#1343

Please check the following:

- [x] All cobrand-specific commits start their commit message with the cobrand in square brackets;
- [x] Have you updated the changelog? If this is not necessary, put square brackets around this: [skip changelog]

Please check the contributing docs, and describe your pull request here.
Screenshots or GIF animations (using e.g. LICEcap) may be helpful.

Please include any issues that are fixed, using "fixes" or "closes" so that
they are auto-closed when the PR is merged.

Thanks for contributing!

### CODE REVIEW

File: web/cobrands/northamptonshire/assets.js [hunk 1]
```diff
@@ -516,4 +516,25 @@ fixmystreet.assets.add($.extend(true, {}, fixmystreet.assets.alloy_defaults, {
     }
 }));
 
+function ncc_check_streetlights() {
+    var relevant_body = OpenLayers.Util.indexOf(fixmystreet.bodies, 'Northamptonshire County Council') > -1;
```
Comment (dracos) [inline web/cobrands/northamptonshire/assets.js]: If you put `body: "Northamptonshire County Council"`  in `northants_defaults`, then you could use `northants_defaults.body` here.

File: web/cobrands/northamptonshire/assets.js [hunk 2]
```diff
@@ -516,4 +516,25 @@ fixmystreet.assets.add($.extend(true, {}, fixmystreet.assets.alloy_defaults, {
...
+        // Either should be shown and already is, or shouldn't be shown and isn't
+        return;
+    }
+
+    if (!relevant) {
+        $('#ncc_streetlights').remove();
+        return;
+    }
+
+    var $msg = $('<p id="ncc_streetlights" class="box-warning">Street lighting in Northamptonshire is maintained by Balfour Beatty on behalf of the County Council under a Street Lighting Private Finance Initiative (PFI) contract. Please view our <a href="https://www3.northamptonshire.gov.uk/councilservices/northamptonshire-highways/roads-and-streets/Pages/street-lighting.aspx">Street Lighting page</a> to report any issues');
```
Comment (dracos) [inline web/cobrands/northamptonshire/assets.js]: Missing an end `</p>`?

File: web/cobrands/northamptonshire/assets.js [hunk 3]
```diff
@@ -618,4 +619,33 @@ function check_emergency() {
...
+        $('#ncc_streetlights').remove();
+        return;
+    }
+
+    var relevant_body = OpenLayers.Util.indexOf(fixmystreet.bodies, northants_defaults.body) > -1;
+    var relevant_cat = $('#form_category').val() == 'Street lighting';
+    var relevant = relevant_body && relevant_cat;
+    var currently_shown = !!$('#ncc_streetlights').length;
+
+  console.log(relevant, currently_shown);
```
Comment (dracos) [inline web/cobrands/northamptonshire/assets.js]: Debug left in

File: web/cobrands/northamptonshire/assets.js [hunk 4]
```diff
@@ -618,4 +619,33 @@ function check_emergency() {
...
+        $('#ncc_streetlights').remove();
+        enable_report_form();
+        return;
+    }
+
+    if (!currently_shown) {
+        var $msg = $('<p id="ncc_streetlights" class="box-warning">Street lighting in Northamptonshire is maintained by Balfour Beatty on behalf of the County Council under a Street Lighting Private Finance Initiative (PFI) contract. Please view our <b><a href="https://www3.northamptonshire.gov.uk/councilservices/northamptonshire-highways/roads-and-streets/Pages/street-lighting.aspx">Street Lighting</a></b> page to report any issues.</p>');
+        $msg.insertBefore('#js-post-category-messages');
+        disable_report_form();
+        hide_responsibility_errors();
```
Comment (dracos) [inline web/cobrands/northamptonshire/assets.js]: Is this line necessary? Isn't it already handled by the removal of the asset/road layer from the category change?

File: web/cobrands/northamptonshire/assets.js [hunk 5]
```diff
@@ -618,4 +619,33 @@ function check_emergency() {
...
+
+    var relevant_body = OpenLayers.Util.indexOf(fixmystreet.bodies, northants_defaults.body) > -1;
+    var relevant_cat = $('#form_category').val() == 'Street lighting';
+    var relevant = relevant_body && relevant_cat;
+    var currently_shown = !!$('#ncc_streetlights').length;
+
+  console.log(relevant, currently_shown);
+    if (!relevant) {
+        $('#ncc_streetlights').remove();
+        enable_report_form();
```
Comment (dracos) [inline web/cobrands/northamptonshire/assets.js]: This worries me due to e.g. changing to a category that would need to hide it again (not on a road or something) but this running after, but as it's all hopefully refactoring I'm not too fussed!

File: web/cobrands/northamptonshire/assets.js [hunk 6]
```diff
@@ -618,4 +619,33 @@ function check_emergency() {
...
+        return;
+    }
+
+    if (!currently_shown) {
```
Comment (dracos) [inline web/cobrands/northamptonshire/assets.js]: Might be nicer (slash easier for me to include in refactor :) ) if it was consistent with the other calls which return first if (relevant === currently_shown). Then this line wouldn't be needed, I think? Unless there was a reason for not doing that.

**Comment by struan** [comment]
Right, I see my error here was trying to take into account the emergency state which is irrelevant as by definition if you've selected street lighting then it's not an emergency. Removing that cleans up the code and removes all the weird edge cases I was seeing.

File: web/cobrands/northamptonshire/assets.js [hunk 7]
```diff
@@ -423,7 +423,7 @@ var northants_defaults = $.extend(true, {}, fixmystreet.assets.alloy_defaults, {
       var emergency_state = ncc_is_emergency_category();
 
       disable_report_form();
-      if (!emergency_state.relevant || emergency_state.body) {
+      if ((!emergency_state.relevant || emergency_state.body) && this.visibility) {
```
Comment (dracos) [inline web/cobrands/northamptonshire/assets.js]: Probably need a similar update to Bucks otherwise the not_found there shows a "Pick an item" message :)

Comment (struan) [reply web/cobrands/northamptonshire/assets.js]: I'm not sure I follow where this is a problem. Can you give me an example?

Comment (dracos) [reply web/cobrands/northamptonshire/assets.js]: I guess it instantly gets removed after, doesn't it. My thought process was (on Bucks) if you switch from e.g. Bollard light not working to non-asset category, `layer_visibilitychanged` is called, which now calls asset_not_found which calls `labeled_defaults.actions.asset_not_found` which sets the message back to the default. It doesn't really need to as it then gets removed, but then that doesn't matter if it does. Hopefully might get tidied up too.

**Comment by dracos** [comment]
@struan I've pushed a fix I think for that that also makes the fn almost identical to road not_found, which makes it feel right?

**Comment by struan** [comment]
Yup, that makes sense.

**Pull request merged**